File: 95. Unique Binary Search Trees II/95. Unique Binary Search Trees II.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  List<TreeNode?> generateTrees(int n) {
    if (n == 0) return [];
    return generateTreesHelper(1, n);
  }
  
  List<TreeNode?> generateTreesHelper(int start, int end) {
    List<TreeNode?> result = [];
    
    if (start > end) {
      result.add(null);
      return result;
    }
    
    for (int i = start; i <= end; i++) {
      List<TreeNode?> leftSubtrees = generateTreesHelper(start, i - 1);
      List<TreeNode?> rightSubtrees = generateTreesHelper(i + 1, end);
      
      for (TreeNode? left in leftSubtrees) {
        for (TreeNode? right in rightSubtrees) {
          TreeNode root = TreeNode(i);
          root.left = left;
          root.right = right;
          result.add(root);
        }
      }
    }
    
    return result;
  }
}

File: 95. Unique Binary Search Trees II/95. Unique Binary Search Trees II.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn generate_trees(n: i32) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
        if n == 0 {
            return vec![];
        }
        Self::generate_trees_helper(1, n)
    }
    
    fn generate_trees_helper(start: i32, end: i32) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
        let mut result = vec![];
        
        if start > end {
            result.push(None);
            return result;
        }
        
        for i in start..=end {
            let left_subtrees = Self::generate_trees_helper(start, i - 1);
            let right_subtrees = Self::generate_trees_helper(i + 1, end);
            
            for left in left_subtrees.iter() {
                for right in right_subtrees.iter() {
                    let mut root = TreeNode::new(i);
                    root.left = left.clone();
                    root.right = right.clone();
                    result.push(Some(Rc::new(RefCell::new(root))));
                }
            }
        }
        
        result
    }
}